contracts/src/chainlink/solidity/emergency/StarkNetValidator.sol

```diff
@@ -74,15 +120,144 @@ contract StarkNetValidator is TypeAndVersionInterface, AggregatorValidatorInterf
+    // NOTICE: we ignore the output of this call (msgHash, nonce).
+    // We also don't raise any events as the 'LogMessageToL2' event will be emitted from the messaging contract.
+    STARKNET_CROSS_DOMAIN_MESSENGER.sendMessageToL2{value: fee}(
+      L2_UPTIME_FEED_ADDR,
+      SELECTOR_STARK_UPDATE_STATUS,
+      payload
+    );
     return true;
   }
+
+  /// @notice L1 oracle is asked for a fast L1 gas price, and the price multiplied by the configured gas estimate
+  function _approximateFee() internal view returns (uint256) {
+    (, int256 fastGasPriceInWei, , , ) = AggregatorV3Interface(s_gasConfig.gasPriceL1Feed).latestRoundData();
+    return (uint256(fastGasPriceInWei) * s_gasConfig.gasEstimate);
+  }
+
+  /**
+   * @notice The selector is the starknet_keccak hash of the function name
+   * @dev StarkNet keccak is defined as the first 250 bits of the Keccak256 hash.
+   *   This is just Keccak256 augmented in order to fit into a field element.
+   * @param fn string function name
+   */
+  function _selectorStarkNet(string memory fn) internal pure returns (uint256) {
+    bytes32 digest = keccak256(abi.encodePacked(fn));
+    return uint256(digest) % 2**250; // get last 250 bits
+  }
+
+  function setGasConfig(uint256 gasEstimate, address gasPriceL1Feed) external onlyOwnerOrConfigAccess {
+    _setGasConfig(gasEstimate, gasPriceL1Feed);
+    emit GasConfigSet(gasEstimate, gasPriceL1Feed);
+  }
+
+  /// @notice internal method that stores the gas configuration
+  function _setGasConfig(uint256 gasEstimate, address gasPriceL1Feed) internal {
+    if (gasPriceL1Feed == address(0)) {
+      revert InvalidGasPriceL1FeedAddress();
+    }
+    s_gasConfig = GasConfig(gasEstimate, gasPriceL1Feed);
+    emit GasConfigSet(gasEstimate, gasPriceL1Feed);
+  }
+
+  /**
+   * @notice sets config AccessControllerInterface contract
+   * @dev only owner can call this
+   * @param accessController new AccessControllerInterface contract address
+   */
+  function setConfigAC(address accessController) external onlyOwner {
+    _setConfigAC(accessController);
+  }
+
+  /// @notice Internal method that stores the configuration access controller
+  function _setConfigAC(address accessController) internal {
+    address previousAccessController = address(s_configAC);
+    if (accessController != previousAccessController) {
+      s_configAC = AccessControllerInterface(accessController);
+      emit ConfigACSet(previousAccessController, accessController);
+    }
+  }
+
+  /**
+   * @notice sets the source aggregator AggregatorInterface contract
+   * @dev only owner can call this
+   * @param source the source aggregator that we'll read data from (on retries)
+   */
+  function setSourceAggregator(address source) external onlyOwner {
+    _setSourceAggregator(source);
+  }
+
+  /// @notice Internal method that sets the source aggregator AggregatorInterface contract
+  function _setSourceAggregator(address source) internal {
```

Comment:
@krebernisak I wonder if we should follow the `propose` and `upgrade` pattern here.